src/meepgeom.cpp

```diff
@@ -307,6 +307,70 @@ bool is_metal(meep::field_type ft, const material_type *material) {
     }
 }
 
+/* implement mirror boundary conditions for i outside 0..n-1: */
+static int mirrorindex(int i, int n) { return i >= n ? 2 * n - 1 - i : (i < 0 ? -1 - i : i); }
+
+meep::vec matgrid_grad(meep::field_type ft, const meep::volume &v, geom_box_tree tp, int oi) {
+  meep::vec gradient(zero_vec(v.dim));
+  vector3 pc = vec_to_vector3(v.center());
+  int x1, y1, z1, x2, y2, z2;
+  meep::realnum dx, dy, dz;
+
+  // assume there is only a single MATERIAL_GRID (i.e., no overlapping grids)
```

Comment:
`material_grids_addgradient_point` should probably throw an error if the grid is not `U_MEAN` or `U_DEFAULT` (or whatever it is called), since it looks like those are currently the only ones supported.

In the `U_MEAN` case, we just need the same type of loop to take the average of the gradient vectors.

Reply:
Done.